rustls/src/quic.rs

```diff
@@ -483,38 +501,51 @@ impl Secrets {
 /// Keys used to communicate in a single direction
 pub struct DirectionalKeys {
     /// Encrypts or decrypts a packet's headers
-    pub header: HeaderProtectionKey,
+    pub header: Box<dyn HeaderProtectionKey>,
     /// Encrypts or decrypts the payload of a packet
-    pub packet: PacketKey,
+    pub packet: Box<dyn PacketKey>,
 }
 
 impl DirectionalKeys {
     pub(crate) fn new(
         suite: &'static Tls13CipherSuite,
-        secret: &hkdf::Prk,
+        secret: &hkdf::OkmOneBlock,
         version: Version,
     ) -> Self {
+        let expander = hkdf::Expander::from_okm(secret, suite.hmac_provider);
         Self {
-            header: HeaderProtectionKey::new(suite, secret, version),
-            packet: PacketKey::new(suite, secret, version),
+            header: suite
+                .quic
+                .header_protection_key(&expander, version),
+            packet: suite
+                .quic
+                .packet_key(suite, &expander, version),
         }
     }
 }
 
-/// A QUIC header protection key
-pub struct HeaderProtectionKey(aead::quic::HeaderProtectionKey);
-
-impl HeaderProtectionKey {
-    fn new(suite: &'static Tls13CipherSuite, secret: &hkdf::Prk, version: Version) -> Self {
-        let alg = match suite.common.bulk {
-            BulkAlgorithm::Aes128Gcm => &aead::quic::AES_128,
-            BulkAlgorithm::Aes256Gcm => &aead::quic::AES_256,
-            BulkAlgorithm::Chacha20Poly1305 => &aead::quic::CHACHA20,
-        };
+/// All AEADs we support have 16-byte tags.
```

Comment:
This seems like a big commit, would be nice to split it up, in particular I'd probably introduce the trait(s) and new code first and then move code around in a separate commit.

Reply:
i have split this up into 3 commits:

- adding the new stuff
- removing the unused BulkAlgorithm
- moving code into crypto::ring